checker/src/org/checkerframework/checker/samelen/SameLenAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,268 @@
+import org.checkerframework.framework.util.AnnotationBuilder;
+import org.checkerframework.framework.util.MultiGraphQualifierHierarchy;
+import org.checkerframework.framework.util.MultiGraphQualifierHierarchy.MultiGraphFactory;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/**
+ * The SameLen Checker is used to determine whether there are multiple arrays in a program that
+ * share the same length. It is part of the Index Checker, and is used as a subchecker by the Index
+ * Checker's components.
+ */
+public class SameLenAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    public static AnnotationMirror UNKNOWN;
+
+    /**
+     * Provides a way to query the Constant Value Checker, which computes the values of expressions
+     * known at compile time (constant prop + folding).
+     */
+    private final ValueAnnotatedTypeFactory valueAnnotatedTypeFactory;
+
+    public SameLenAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        valueAnnotatedTypeFactory = getTypeFactoryOfSubchecker(ValueChecker.class);
+        UNKNOWN = AnnotationUtils.fromClass(elements, SameLenUnknown.class);
+        this.postInit();
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
+        return new SameLenQualifierHierarchy(factory);
+    }
+
+    /**
+     * Checks whether the two string arrays contain at least one string that's the same. Not a smart
+     * algorithm; meant to be run over small sets of data.
+     *
+     * @param a1 the first string array
+     * @param a2 the second string array
+     * @return true if there is the intersection is non-empty; false otherwise
+     */
+    private boolean overlap(String[] a1, String[] a2) {
+        for (String a : a1) {
+            for (String b : a2) {
+                if (a.equals(b)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This function finds the union of the values of two annotations. Both annotations must have a
+     * value field; otherwise the function will fail.
+     *
+     * @param a1 an annotation with a value field
+     * @param a2 an annotation with a value field
+     * @return the set union of the two value fields
+     */
+    private AnnotationMirror getCombinedSameLen(String[] a1, String[] a2) {
+
+        List<String> a1Names = new ArrayList<>();
+        List<String> a2Names = new ArrayList<>();
+
+        for (String s : a1) {
+            a1Names.add(s);
+        }
+        for (String s : a2) {
+            a2Names.add(s);
+        }
+
+        HashSet<String> newValues = new HashSet<String>(a1Names.size() + a2Names.size());
+
+        newValues.addAll(a1Names);
+        newValues.addAll(a2Names);
+        Object[] values = newValues.toArray();
+        String[] names = new String[values.length];
+        for (int i = 0; i < names.length; i++) {
+            names[i] = values[i].toString();
```

Comment:
Use the https://docs.oracle.com/javase/7/docs/api/java/util/List.html#toArray(T[]) rather than this loop.